x-pack/plugins/alerting/server/lib/get_create_task_runner_function.ts

```diff
@@ -0,0 +1,106 @@
+import { ActionsPlugin } from '../../../actions';
+import { AlertType, Services, AlertServices } from '../types';
+import { TaskInstance } from '../../../task_manager';
+import { createFireHandler } from './create_fire_handler';
+import { createAlertInstanceFactory } from './create_alert_instance_factory';
+import { AlertInstance } from './alert_instance';
+import { getNextRunAt } from './get_next_run_at';
+import { validateAlertTypeParams } from './validate_alert_type_params';
+
+interface CreateTaskRunnerFunctionOptions {
+  getServices: (basePath: string) => Services;
+  alertType: AlertType;
+  fireAction: ActionsPlugin['fire'];
+  internalSavedObjectsRepository: SavedObjectsClientContract;
+}
+
+interface TaskRunnerOptions {
+  taskInstance: TaskInstance;
+}
+
+export function getCreateTaskRunnerFunction({
+  getServices,
+  alertType,
+  fireAction,
+  internalSavedObjectsRepository,
+}: CreateTaskRunnerFunctionOptions) {
+  return ({ taskInstance }: TaskRunnerOptions) => {
+    return {
+      run: async () => {
+        const alertSavedObject = await internalSavedObjectsRepository.get(
+          'alert',
+          taskInstance.params.alertId
+        );
+
+        // Validate
+        const validatedAlertTypeParams = validateAlertTypeParams(
+          alertType,
+          alertSavedObject.attributes.alertTypeParams
+        );
+
+        const fireHandler = createFireHandler({
+          alertSavedObject,
+          fireAction,
+          basePath: taskInstance.params.basePath,
+        });
+        const alertInstances: Record<string, AlertInstance> = {};
+        const alertInstancesData = taskInstance.state.alertInstances || {};
+        for (const id of Object.keys(alertInstancesData)) {
+          alertInstances[id] = new AlertInstance(alertInstancesData[id]);
+        }
+        const alertInstanceFactory = createAlertInstanceFactory(alertInstances);
+
+        const alertTypeServices: AlertServices = {
+          ...getServices(taskInstance.params.basePath),
+          alertInstanceFactory,
+        };
+
+        const alertTypeState = await alertType.execute({
+          services: alertTypeServices,
+          params: validatedAlertTypeParams,
+          state: taskInstance.state.alertTypeState || {},
+          scheduledRunAt: taskInstance.state.scheduledRunAt,
+          previousScheduledRunAt: taskInstance.state.previousScheduledRunAt,
+        });
+
+        for (const alertInstanceId of Object.keys(alertInstances)) {
+          const alertInstance = alertInstances[alertInstanceId];
+
+          // Unpersist any alert instances that were not explicitly fired in this alert execution
+          if (!alertInstance.shouldFire()) {
+            delete alertInstances[alertInstanceId];
+            continue;
+          }
+
+          const { actionGroup, context, state } = alertInstance.getFireOptions()!;
+          await fireHandler(actionGroup, context, state);
```

Comment:
I think `for..of` with the `await` will hold up the loop on each action?  If there are potentially a lot of alert instances and the handler is a potentially long operation like sending an email, this might take a long time to complete.  

If you used forEach instead we could take action on all the instances at the same time?